### PR TITLE
Upgrade Ear plugin

### PR DESCRIPTION
**Pull request by asodja**
Except `DeploymentDescriptor.getFileName()`, since it has weird setter

### CODE REVIEW

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test this

File: platforms/jvm/ear/src/main/java/org/gradle/plugins/ear/Ear.java [hunk 1]
```diff
@@ -108,10 +103,15 @@ public Ear() {
...
-                        // (for instance, due to modules discovered)
-                        descriptor.writeTo(new OutputStreamWriter(outputStream))
+                    action(outputStream -> {
+                            // delay obtaining contents to account for descriptor changes
+                            // (for instance, due to modules discovered)
+                            // We copy the descriptor since properties are already finalized at that moment with cc
+                            DeploymentDescriptor copyDescriptor = getObjectFactory().newInstance(DefaultDeploymentDescriptor.class).copyFrom(descriptor);
+                            copyDescriptor.getModules().addAll(topLevelModules.get());
+                            copyDescriptor.getLibraryDirectory().set(descriptor.getLibraryDirectory().orElse(getLibDirName()));
+                            copyDescriptor.writeTo(new OutputStreamWriter(outputStream));
```
Comment (asodja) [inline platforms/jvm/ear/src/main/java/org/gradle/plugins/ear/Ear.java:113]: 💭 We do this copy logic because properties can be finalized at that moment with cc

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle clean

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test this

File: platforms/jvm/ear/build.gradle.kts [hunk 2]
```diff
@@ -14,17 +14,16 @@ errorprone {
...
+    api(projects.core)
     api(projects.coreApi)
     api(projects.languageJvm)
     api(projects.modelCore)
     api(projects.platformJvm)
 
     implementation(projects.serviceLookup)
     implementation(projects.stdlibJavaExtensions)
-    implementation(projects.core)
     implementation(projects.fileOperations)
```
Comment (lptr) [inline platforms/jvm/ear/build.gradle.kts:27]: Why is this necessary?

File: platforms/jvm/ear/src/main/java/org/gradle/plugins/ear/Ear.java (hunk 1 above)
Comment (lptr) [reply platforms/jvm/ear/src/main/java/org/gradle/plugins/ear/Ear.java:113]: :x: Why is it a problem that the properties are finalized? IIUC, this is about having to call `setLibraryDirectory()`, right? Why do we have to call this at execution time? Can't we set that as a convention instead?

File: platforms/jvm/ear/src/main/java/org/gradle/plugins/ear/Ear.java [hunk 3]
```diff
@@ -130,19 +137,24 @@ private OutputChangeListener outputChangeListener() {
         return getServices().get(OutputChangeListener.class);
     }
 
-    private void recordTopLevelModules(FileCopyDetails details) {
+    private SetProperty<EarModule> recordTopLevelModules(FileCopyDetails details) {
```
Comment (lptr) [inline platforms/jvm/ear/src/main/java/org/gradle/plugins/ear/Ear.java:140]: :x: Why return a `SetProperty` here instead of a `Set`?

File: platforms/jvm/ear/src/main/java/org/gradle/plugins/ear/Ear.java [hunk 4]
```diff
@@ -60,24 +61,22 @@
...
-        lib = getRootSpec().addChildBeforeSpec(getMainSpec()).into(
-            callable(() -> GUtil.elvis(getLibDirName(), DEFAULT_LIB_DIR_NAME))
-        );
+        lib = getRootSpec().addChildBeforeSpec(getMainSpec()).into(getLibDirName().orElse(DEFAULT_LIB_DIR_NAME));
+        AtomicReference<SetProperty<EarModule>> topLevelModules = new AtomicReference<>(getObjectFactory().setProperty(EarModule.class));
         getMainSpec().appendCachingSafeCopyAction(action(details -> {
             if (generateDeploymentDescriptor.get()) {
                 checkIfShouldGenerateDeploymentDescriptor(details);
-                recordTopLevelModules(details);
+                topLevelModules.set(union(topLevelModules.get(), recordTopLevelModules(details)));
```
Comment (lptr) [inline platforms/jvm/ear/src/main/java/org/gradle/plugins/ear/Ear.java:79]: This only needs to be a `Provider<Set>`, not a `SetProperty` necessarily. Can we use `.zip()` here instead? Or can we just use `Set` and union that? I don't understand why we are using `Property` types here at all.

File: platforms/jvm/ear/src/main/java/org/gradle/plugins/ear/EarPlugin.java [hunk 5]
```diff
@@ -123,12 +123,8 @@ private void setupEarTask(final Project project, EarPluginConvention convention,
...
         DeploymentDescriptor deploymentDescriptor =  DeprecationLogger.whileDisabled(() -> convention.getDeploymentDescriptor());
         if (deploymentDescriptor != null) {
-            if (deploymentDescriptor.getDisplayName() == null) {
-                deploymentDescriptor.setDisplayName(project.getName());
-            }
-            if (deploymentDescriptor.getDescription() == null) {
-                deploymentDescriptor.setDescription(project.getDescription());
-            }
+            deploymentDescriptor.getDisplayName().convention(project.getName());
+            deploymentDescriptor.getDescription().convention(project.getDescription());
```
Comment (lptr) [inline platforms/jvm/ear/src/main/java/org/gradle/plugins/ear/EarPlugin.java:127]: 💭 This would work so much nicer if the `project` properties were lazy, too...

File: platforms/jvm/ear/src/main/java/org/gradle/plugins/ear/descriptor/internal/DefaultDeploymentDescriptor.java [hunk 6]
```diff
@@ -343,6 +276,26 @@ public void execute(XmlProvider xmlProvider) {
...
+        return module;
+    }
+
+    private DefaultEarSecurityRole newDefaultEarSecurityRole(String roleName, String description) {
+        DefaultEarSecurityRole role = objectFactory.newInstance(DefaultEarSecurityRole.class);
+        role.getRoleName().set(roleName);
+        role.getDescription().set(description);
+        return role;
+    }
+
```
Comment (lptr) [inline platforms/jvm/ear/src/main/java/org/gradle/plugins/ear/descriptor/internal/DefaultDeploymentDescriptor.java:298]: ❓ This looks fishy. Shouldn't this be handled by a `DomainObjectContainer` or something instead?

File: platforms/jvm/ear/src/main/java/org/gradle/plugins/ear/descriptor/internal/DefaultDeploymentDescriptor.java [hunk 7]
```diff
@@ -451,4 +404,18 @@ private Object nodeNameFor(String name) {
...
+        getApplicationName().set(descriptor.getApplicationName());
+        getInitializeInOrder().set(descriptor.getInitializeInOrder());
+        getDescription().set(descriptor.getDescription().getOrNull());
+        getDisplayName().set(descriptor.getDisplayName().getOrNull());
+        getLibraryDirectory().set(descriptor.getLibraryDirectory());
+        getModules().set(descriptor.getModules());
+        getSecurityRoles().set(descriptor.getSecurityRoles());
+        getModuleTypeMappings().set(descriptor.getModuleTypeMappings());
+        return this;
+    }
```
Comment (lptr) [inline platforms/jvm/ear/src/main/java/org/gradle/plugins/ear/descriptor/internal/DefaultDeploymentDescriptor.java:420]: :x: This looks like part of an anti-pattern, and would be good if we could delete.

File: platforms/jvm/ear/src/test/groovy/org/gradle/plugins/ear/EarTest.groovy [hunk 8]
```diff
@@ -21,13 +21,30 @@ import org.gradle.api.file.CopySpec
...
+        @Override
+        PathToFileResolver newResolver(File baseDir) {
+            return null
+        }
+
+        @Override
+        boolean canResolveRelativePath() {
+            return false
+        }
+    }
```
Comment (lptr) [inline platforms/jvm/ear/src/test/groovy/org/gradle/plugins/ear/EarTest.groovy:47]: ❓ Can we use a mock here instead?

File: platforms/jvm/ear/src/main/java/org/gradle/plugins/ear/Ear.java (hunk 1 above)
Comment (asodja) [reply platforms/jvm/ear/src/main/java/org/gradle/plugins/ear/Ear.java:113]: > Why is it a problem that the properties are finalized?

The main problem why we do it here is because we want to add these top level modules, so this line:
```
copyDescriptor.getModules().addAll(topLevelModules.get())
``` 

I have to refresh my memory why `setLibraryDirectory()` is a problem here

File: platforms/jvm/ear/build.gradle.kts (hunk 2 above)
Comment (asodja) [reply platforms/jvm/ear/build.gradle.kts]: Reverted that. Project health showed this as a problem, but it seems it's not a problem anymore after rebase

File: platforms/jvm/ear/src/main/java/org/gradle/plugins/ear/Ear.java (hunk 3 above)
Comment (asodja) [reply platforms/jvm/ear/src/main/java/org/gradle/plugins/ear/Ear.java]: Yes, we can use `Set`, changed to `Set`

File: platforms/jvm/ear/src/main/java/org/gradle/plugins/ear/Ear.java (hunk 4 above)
Comment (asodja) [reply platforms/jvm/ear/src/main/java/org/gradle/plugins/ear/Ear.java]: > Or can we just use Set and union that?

We now use a Set

File: platforms/jvm/ear/src/main/java/org/gradle/plugins/ear/descriptor/internal/DefaultDeploymentDescriptor.java (hunk 6 above)
Comment (asodja) [reply platforms/jvm/ear/src/main/java/org/gradle/plugins/ear/descriptor/internal/DefaultDeploymentDescriptor.java:298]: This are basically just a helper methods to set properties. I don't think `DomainObjectContainer` handles this, since we don't inject values to `constructor` but we set properties via `.set()`

File: platforms/jvm/ear/src/test/groovy/org/gradle/plugins/ear/EarTest.groovy (hunk 8 above)
Comment (asodja) [reply platforms/jvm/ear/src/test/groovy/org/gradle/plugins/ear/EarTest.groovy]: Yes, we can mock/stub it. We use a `Stub(PathToFileResolver )` now

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test this

File: platforms/jvm/ear/src/main/java/org/gradle/plugins/ear/descriptor/internal/DefaultDeploymentDescriptor.java (hunk 7 above)
Comment (asodja) [reply platforms/jvm/ear/src/main/java/org/gradle/plugins/ear/descriptor/internal/DefaultDeploymentDescriptor.java]: I was able to avoid that with this change: [44e4b5a](https://github.com/gradle/gradle/pull/30496/commits/44e4b5aa3a2fc59c2bcd4b28541cc100a30336cf)

But I think we should merge that directly to master to have smaller changes on EAP branch.

**Comment by asodja** [comment]
@bot-gradle help

**Comment by asodja** [comment]
@bot-gradle clean

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle clean

File: platforms/jvm/ear/src/main/java/org/gradle/plugins/ear/descriptor/internal/DefaultDeploymentDescriptor.java (hunk 6 above)
Comment (lptr) [reply platforms/jvm/ear/src/main/java/org/gradle/plugins/ear/descriptor/internal/DefaultDeploymentDescriptor.java:283]: Is it an option though to inject via constructor parameters, and use a `DOC`? I am always wary about naked use of `ObjectFactory` or any other direct instantiation. These cases should be handled via regualr injection.

File: platforms/jvm/ear/src/main/java/org/gradle/plugins/ear/Ear.java [hunk 9]
```diff
@@ -199,7 +206,7 @@ private DeploymentDescriptor forceDeploymentDescriptor() {
      * A location for dependency libraries to include in the 'lib' directory of the EAR archive.
      */
     @Internal
-    @ToBeReplacedByLazyProperty(comment = "Should this be lazy?")
+    @NotToBeReplacedByLazyProperty(because = "Nested property")
```
Comment (lptr) [inline platforms/jvm/ear/src/main/java/org/gradle/plugins/ear/Ear.java:209]: ❓ If this is a nested property, shouldn't it be marked as `@Nested`? I know `CopySpec`s are handled in a special way, but maybe we can find a better explanation here?

File: platforms/jvm/ear/src/main/java/org/gradle/plugins/ear/Ear.java [hunk 10]
```diff
@@ -260,7 +260,7 @@ public Property<Boolean> getGenerateDeploymentDescriptor() {
      * The deployment descriptor configuration.
      */
     @Internal
-    @ToBeReplacedByLazyProperty
+    @NotToBeReplacedByLazyProperty(because = "Nested property")
```
Comment (lptr) [inline platforms/jvm/ear/src/main/java/org/gradle/plugins/ear/Ear.java:263]: :x: This is not a nested property, so let's not give that as the reason here.

Comment (asodja) [reply platforms/jvm/ear/src/main/java/org/gradle/plugins/ear/Ear.java]: Yes this property is not annotated with `@Nested` but It looks like a nested property just annotated as `@Internal`, I guess so we don't snapshot its values.

Should we migrate this one or what do you think would be a reason here?

Comment (asodja) [reply platforms/jvm/ear/src/main/java/org/gradle/plugins/ear/Ear.java]: I changed it to `Nested like property`, not sure what we should put otherwise

File: platforms/jvm/ear/src/main/java/org/gradle/plugins/ear/Ear.java (hunk 9 above)
Comment (asodja) [reply platforms/jvm/ear/src/main/java/org/gradle/plugins/ear/Ear.java]: I changed to `Read-only CopySpec property`.

I don't think this should be lazy or either `@Nested` since it's CopySpec. 🤔

File: platforms/jvm/ear/src/main/java/org/gradle/plugins/ear/Ear.java (hunk 10 above)
Comment (asodja) [reply platforms/jvm/ear/src/main/java/org/gradle/plugins/ear/Ear.java]: Optionally we can also try to annotate it as `@Nested` and annotate all properties as `@Internal` if that is any better

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test RFN
